File: today3/evaluate-reverse-polish-notation.ts
/**
 * https://leetcode.com/problems/evaluate-reverse-polish-notation/description/
 */
export { evalRPN };

function evalRPN(tokens: string[]): number {
  const stack: number[] = [];

  for (let i = 0; i < tokens.length; i++) {
    if (tokens[i] === '+' || tokens[i] === '-' || tokens[i] === '*' || tokens[i] === '/') {
      const right = stack.pop() as number;
      const left = stack.pop() as number;
      let result: number;

      switch (tokens[i]) {
        case '+':
          result = left + right;
          break;
        case '-':
          result = left - right;
          break;
        case '*':
          result = left * right;
          break;
        case '/':
          const divided = left / right;
          if (divided < 0) result = Math.ceil(divided);
          else result = Math.floor(divided);
          break;
      }
      stack.push(result!);
    } else {
      const parsed = Number.parseInt(tokens[i]);
      stack.push(parsed);
    }
  }

  return stack[0];
}

File: returned/python/today12/evaluate_reverse_polish_notation.py
# https://leetcode.com/problems/evaluate-reverse-polish-notation/description/
from typing import List

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for t in tokens:
            if t == '+':
                stack.append(stack.pop() + stack.pop())
            elif t == '-':
                a, b = stack.pop(), stack.pop()
                stack.append(b-a)
            elif t == '*':
                stack.append(stack.pop() * stack.pop())
            elif t == '/':
                a, b = stack.pop(), stack.pop() 
                stack.append(int(b/a))
            else:
                stack.append(int(t))
        return stack[0]